ios/Codeforces Watcher/Application/features/contests/ContestsNavigationBar.swift

```diff
@@ -0,0 +1,121 @@
+import SwiftUI
+
+struct ContestsNavigationBar: View {
+    
+    var filterItems: [ContestFilterView.UIModel]
+    
+    private let filtersViewAnimationDuration = 0.3
+    @State private var isContestFiltersView = false
+    @State private var geometryHeight: CGFloat = 0
+    
+    var body: some View {
+        HStack {
+            CommonText((isContestFiltersView ? "filters" : "Contests").localized)
+                .font(.headerMedium)
+                .foregroundColor(Palette.black.swiftUIColor)
+                .animation(nil)
+                .modifier(Shake(animatableData: isContestFiltersView ? 1 : 0))
+            
+            Spacer()
+            
+            if isContestFiltersView {
+                Button(action: {
+                    withAnimation(.easeOut(duration: filtersViewAnimationDuration)) {
+                        withAnimation(.spring(response: 0.3, dampingFraction: 0.7)) {
+                            isContestFiltersView.toggle()
+                        }
+                    }
+                }, label: {
+                    Image("crossIconNew")
+                        .renderingMode(.original)
+                })
+            } else {
+                Button(action: {
+                    withAnimation(.easeOut(duration: filtersViewAnimationDuration)) {
+                        withAnimation(.spring(response: 0.3, dampingFraction: 0.7)) {
+                            isContestFiltersView.toggle()
+                        }
+                    }
+                }, label: {
+                    Image("filterIcon")
+                        .renderingMode(.original)
+                })
+            }
+        }
+        .frame(maxWidth: .infinity)
+        .frame(height: 56, alignment: .center)
+        .padding(.horizontal, 20)
+        
+        if isContestFiltersView {
+            ContestFiltersView
+                .padding([.horizontal, .vertical], 20)
+        }
+    }
+    
+    @ViewBuilder
+    private var ContestFiltersView: some View {
```

Comment:
Wow, it looks complicated and elegant at the same time 👍

Reply:
I'd slightly prefer to have `struct ... : View` for sub-components as they allow to extract more details. For example, here it could be constants you are using - 50 and 70, which aren't clear at the moment (I mean giving them good names would improve readability).

Reply:
Constants can also be inside private components, or do you mean that it would be better to extract this component as a struct anyway?